### ISSUE TITLE
rundeck "Kill Job" does not terminate tree of child processes

### ISSUE THREAD

**Issue by lruslan** [rundeck/rundeck#1038]
It seem rundeck "Kill Job" logic does not ensure that child processes killed and exited, leaving child process running in background in detached state.

Environment:
Rundeck version: latest compiled from main 2.4.0-0-SNAPSHOT also official release 2.2.3-1
OS:Ubuntu 14.04.1 LTS , Red Hat Enterprise Linux Server release 6.5

Use case:
Job called locally or remotely ( using script node executor because sometimes we need pseudo terminal to be created ssh -q -tt ${node.username}@${node.hostname} ${exec.command})
Yaml job definition:

```
- project: test
  sequence:
    keepgoing: false
    strategy: node-first
    commands:
    - exec: /tmp/1.sh
  name: job1
```

Script 1.sh

```
#!/bin/bash
python /tmp/1.py
```

Script /tmp/1.py

```
import time
while 1 > 0:
    time.sleep(5)
    print "Hi"
```

Actual behavior:
After job is started press "Kill job" .
Job terminates with log entry:

```
Result: 2147483647
Failed: NonZeroResultCode: Result code was 2147483647
Execution failed: 28: [Workflow step failures: {1=Dispatch failed on 1 nodes: [localhost: NonZeroResultCode: Result code was 2147483647]}, Node failures: {localhost=[NonZeroResultCode: Result code was 2147483647]}]
```

ps axu, and pstree display process still alive and running:

```
#ps ax
rundeck   5386  0.0  0.0  17976  1432 ?        S    12:33   0:00 /bin/bash /tmp/1.sh
rundeck   5387  0.0  0.1  29840  4996 ?        S    12:33   0:00 python /tmp/1.py

~# pstree -a
init
  ├─1.sh /tmp/1.sh
  │   └─python /tmp/1.py
```

Similar problem with script node executor : 
Executor script:

```
ssh -q -tt ${node.username}@${node.hostname} ${exec.command}  
```

After kill pressed rundeck reports job as aborted
Log shows message:

```
java.lang.InterruptedException
Failed: Unknown: java.lang.InterruptedException
Execution failed: 7: [Workflow step failures: {1=Dispatch failed on 1 nodes: [docker: Unknown: java.lang.InterruptedException]}, Node failures: {docker=[Unknown: java.lang.InterruptedException]}]
```

pstree still reports running process 

```
  ├─java -Djava.security.auth.login.config=/etc/rundeck/jaas-loginmodule.conf -Dloginmodule.name=RDpropertyfilelogin-Drdeck.config=/etc/rund
  │   ├─ssh -q -oStrictHostKeyChecking=no -tt root@172.17.0.2 /tmp/1417433949315-docker-dispatch-script.tmp.sh
  │   └─43*[{java}]

```

Expected behavior:
All process started by job terminated or job termination aborted if processes failed to stop, or at list show message about job killed but child process left running.

It seem rundeck use quite simplistic approach to terminate job thread and does not provide any logic to handle killing of tree of child processes. For example jenkins has much more advanced logic to track process tree and handle killing: https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/util/ProcessTree.java

**Comment by mathpl** [comment]
+1, I would like this. Could you explain how it's currently implemented?

The behavior I'm seeing: rundeck creates another connection when I try to kill a job and runs "kill -s 15" on it, which isn't a valid command since it's missing a pid. Then I get the error on rundeck side as described by lruslan.

**Comment by jankei83** [comment]
+1

In 2.5.1 there is still no fix for this problem.

**Comment by mtrofimm** [comment]
+1

**Comment by blairpaine** [comment]
+1

**Comment by steveblackmon** [comment]
+1

**Comment by gtmx** [comment]
+1

**Comment by gMickey** [comment]
+1
Custom terminate handler would be great.

**Comment by jedinerd** [comment]
+1

**Comment by guillaumevoisin** [comment]
Are we considering doing it in the near future?

**Comment by mtrofimm** [comment]
actually i found a kind of workaround to make the job a direct child process of the rundeck server process (without 'sh' process in the middle)
In the "Script file or URL" step, you can use 'exec' in the "Invocation String" field. 
I use such command for symfony2 php console commands in the following way:
File Path/URL: /usr/bin/php
Arguments: path_to_console command_params
Invocation String: exec

**Comment by purduemike** [comment]
This seems very rudimentary.  Are you guys giving this any consideration (timeline)?

**Comment by scopenco** [comment]
I found another workaround. 

1) Create a script like `/usr/bin/exec.sh` with content
```
#!/bin/sh
exec $*
```
2) In the "Script file or URL" step i set:
* `File Path/URL` - `/usr/bin/exec.sh`
* `Arguments` - my build script with with args

And `Kill jobs` will kill parent with childs.

**Comment by mtrofimm** [comment]
@scopenco it's actually the same workaround as mine :)

**Comment by EgleGrublyte** [comment]
+1

**Comment by alv000h** [comment]
+1

@gschueler any news on this topic?

**Comment by zolech** [comment]
+1

Right now I have to restart rundeck once in a while because of non killable jobs that are running forever

**Comment by mtrofimm** [comment]
@zolech did you try my workaround from https://github.com/rundeck/rundeck/issues/1038#issuecomment-232064537 ?

**Comment by zolech** [comment]
@mtrofimm yes, we've changed one job as you described but we need some time to confirm that it works. If it works (I really hope so) we will go with the rest of them (we are also using symfony2 cli commands so it is similar). But rundeck needs proper solution for this so I voted for it anyway.

**Comment by jvrobert** [comment]
+1

This seems rather important, job management is a critical piece of a job execution/scheduling engine.  I can't get 'kill' to work at all for any process, even the parent/top level job isn't properly handled.  I'd be happy if just my top level process got killed.

I poked around a bit seeing if I could just shove a Process.destroy() in there somewhere but realized it's all a bit complicated.  I think one single way to execute a process should used, not e.g. Ant ExecTask (which is essentially uncontrollable) in one situation, and a direct runtime Process in another.

It would be a fair amount of work but seems the "right way" to do this is via JNI with full blown process management, but even a two-stage kill (TERM then KILL) executed by Runtime.exec would be great.

**Comment by zsoca8711** [comment]
+1

**Comment by gschueler** [comment]
this change was added in 2.8.1: https://github.com/rundeck/rundeck/pull/2433

**Comment by DiegoQueiroz** [comment]
Is there any updates on this? There are, at least, 10 other notified issues that can be merged with this one, but apparently, nobody is working on this.

**Comment by Kronnork** [comment]
Has it been tested in 2.8.3 release?

**Comment by DiegoQueiroz** [comment]
Yes. Same behavior.

**Comment by Hackeruncle** [comment]
+1

In 2.9.3 there is still no fix for this problem.

```
Result: 2147483647
--
08:05:00 |   |   | Failed: NonZeroResultCode: Result code was 2147483647
08:05:00 |   |   | java.util.concurrent.CancellationException
08:05:00 |   |   | at java.util.concurrent.FutureTask.report(FutureTask.java:121)
08:05:00 |   |   | at java.util.concurrent.FutureTask.get(FutureTask.java:192)
08:05:00 |   |   | at com.google.common.util.concurrent.Uninterruptibles.getUninterruptibly(Uninterruptibles.java:137)
08:05:00 |   |   | at com.google.common.util.concurrent.Futures$6.run(Futures.java:1308)
08:05:00 |   |   | at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
08:05:00 |   |   | at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
08:05:00 |   |   | at java.lang.Thread.run(Thread.java:745)
08:05:00 |   |   | Cancellation while running step [1]
```

**Comment by lndevnull** [comment]
+1
I'd like to see this feature working for commands over SSH as well.   One
possible work-around for me would be to pass the Timeout option a script
wrapper which wraps a sub script or set of commands within GNU coreutils'
timeout which should send a local kill signal to the sub-processes if still
running.

**Comment by zsoca8711** [comment]
+1 Like  @lndevnull , we also  would really love this for commands over SSH.

**Comment by linux2647** [comment]
I've noticed that if a command is running and it is killed from Rundeck, while it may not immediately be killed, if it tries to write to stdout or stderr, the fds have been closed, so an error will occur, bringing down the entire execution tree.

So,
```
bash -c 'while true; do sleep 1; done'
```
won't be killed by the _Kill Job_ button, but
```
bash -c 'while true; do echo "hello world"; sleep 1; done'
```
will, but not until it tries to `echo`.

**Comment by dpomnean** [comment]
+1 We are seeing the same issue as well. We are having servers reboot when they aren't supposed to, because scripts aren't being killed.

**Comment by roman-io** [comment]
+1

**Comment by abroskin** [comment]
Hi all,
@scopenco, the fix you proposed somehow doesn't work for me. The child process is there even after the parent is killed.
What seems to work in my case is adding a standard Linux tool "timeout" before the command you run.
It's kind of introducing the second execution time limit, but even if it is set to a bigger time interval, the first rundeck's timeout starts working as expected and killing all child processes. As well as manual Kill Job button.
Did somebody try this approach or I miss something?

**Comment by thancock14** [comment]
I also experienced this problem when executing a c# exe using Powershell's Start-Process in the "Script - Execute an inline script" step in a Job.  

`Start-Process -Wait -NoNewWindow -FilePath "E:\Working\${job.group}\${job.name}\ETL.exe" -ArgumentList "arg1 arg2" -WorkingDirectory "E:\working\${job.group}\${job.name}"
exit $lastexitcode`

but that had the same problem as those who mentioned before.
so this would be hugely beneficial to fix.
But I was able to use the workaround that @DiegoQueiroz put forward with the RunMe.py
I adjusted the RunMe.py to not step through a .bat file, but rather to chdir to working directory and simply call a c# executable. that worked great!  
thanks @DiegoQueiroz !

**Comment by dandunckelman** [comment]
In my case, I'm running an "adhoc script URL" via `POST /api/14/project/[PROJECT]run/url`.

It's a bash script that downloads/executes a python script, which runs `terraform apply` and manages its output.

When we abort that job, a lot of terraform provisioners are left running (and those are looping waiting for certain events to occur...because the job was aborted, so those events will never happen). I'm hoping to find a way to solve that to avoid lots of orphaned processes that never die.

**Comment by DiegoQueiroz** [comment]
Yes. It is still relevant.

It affects mostly Windows, so this probably is why people are not freaking out in the comments.

Besides that, my hack and @thancock14 worked around the problem while there isn't a definitive solution.

**Comment by geekguy** [comment]
This issue is still there.

Is there any workaround for now?

**Comment by ProTip** [comment]
Hi @geekguy 

This is a rather complicated topic at the system level, and I believe this link provides a good overview of some options available: https://unix.stackexchange.com/questions/124127/kill-all-descendant-processes .

That being said, for Linux here are some suggestions:
## SSH
For commands running over `ssh` I might recommend forcing `tty` when possible. `SIGHUP` is sent to session members only when the leader is disconnected from the `tty`. So, by forcing `tty` you increase the likelihood that a job timeout, network issue, or other event that would cause the parent command to exit without waiting on children will result in the tree being killed as well.

## Screen
Locally executing the entry-point via `screen` will also result in a `SIGHUP` being sent. As an expiriment place the following in `sleep.sh`:
```
#!/bin/bash

(sleep 10) &
```

Execute this and observe that the sleep command is still running:
```
bash sleep.sh && ps axf | grep sleep
```

Execute this same command through screen and notice the sleep command gets killed when the process and screen exits:
```
screen bash sleep.sh && ps axf | grep sleep
```

## Summary
I believe there is certainly room for adding more advanced execution options that can reliably track all forks and cleanup, however these techniques may offer an adequate workaround in the meantime. Unfortunately I'm pretty foggy on this subject as it pertain to Windows and hope to dive into it further.

**Comment by alv000h** [comment]
@gschueler is this fixed in 3?

**Comment by dandunckelman** [comment]
@alv000h no. I'm still seeing this on v3.1.0-20190731.

**Comment by tsolodov** [comment]
found solution for myself: add error handler -> run jobs that does kill process, before that you need to export pid of running process.

**Comment by yogeshkumark** [comment]
+1, still facing the same issue. 
Could we able to set the timeout for step level ? , that could solve the problem ?

**Comment by falloutphil** [comment]
+1 - still an issue in Linux on 3.1.2-20190927 - would love to see a proper build-in solution.  It's probably the biggest missing fundamental/practical feature in Rundeck for my use.

Having had some servers fill up with stale jobs - I reduced the issue to a Rundeck command that remotely (via SSH) calls a simple bash script that echos its own pid and then goes to sleep for 2 minutes.  This can be shown to persist after killing the job in Rundeck. 

Forcing tty seems to work - but it's a bit of a hack.  Timeouts are too crude when jobs take significantly variable times.  Also, FYI I couldn't get this fix to work on 3.1.2: https://github.com/rundeck/rundeck/issues/1038#issuecomment-232064537 - the behaviour was the same with the remote script continuing after killing the job.

**Comment by svenbs** [comment]
+1

**Comment by rshingleton** [comment]
+1

Why is this still an issue **_### 7 years later_**? I'm running into zombie processes from custom script plugins that fork processes. When the Rundeck job is killed there doesn't appear to be a way to catch/handle the termination. Seems like there should be a way to send a signal down the line.

**Comment by ezlo-codrin** [comment]
This is still an issue. Can it be fixed?

**Comment by yogeshkumark** [comment]
Some rare cases this could happen. if it is happening very frequently.
there could be 2 reasons
1) validate the network bandwidth or increase the network bandwidth at the
system level.
2) validate network transfer loss and receiver loss. if there is some value
then you need to fix the network only.

YOGESH KUMAR
(09884343891)

A computer is like air conditioning: it becomes useless when you open
windows.
<-Fighting 4 Freedom->

**Comment by jipipayo** [comment]
I suggest to remove the button "kill job" from the ui . That will avoid to not "lie" to the user, and left unhandled zombie jobs running out of rundeck control and giving the false impression that this job is successfully terminated.

**Comment by RadicalEdd** [comment]
+1
It has been 8 Years and we still don't have this one implemented?

**Comment by zachzoref** [comment]
This is still an issue, 9 years later...
guys, please help 😿

**Comment by mgravante** [comment]
entering tenth year..

**Comment by LevonW-IIS** [comment]
+1

**Comment by tabortrailways** [comment]
+1 not that it means anything 10 years in

**Comment by fdevans** [comment]
Thanks for the report. As part of a cleanup of our issue tracker, we're closing issues older than 12 months, in line with Rundeck's support policy of one year from a release's date (see the [release calendar](https://docs.rundeck.com/docs/history/release-calendar.html) for currently supported versions). This isn't a judgment that the report is invalid — many of these are likely still real — but we can't verify against a version this old.

If you're still seeing this on a currently supported release (6.1.x or 5.20.x+), please open a new issue using our [bug report template](https://github.com/rundeck/rundeck/issues/new?template=bug_report.yml) with your version and reproduction steps. Please don't reopen this issue — a fresh issue keeps things tracked against a supported version and gets it in front of us properly. We're actively working on improving how issues get submitted and triaged going forward.

Thanks for your patience,
The Rundeck Team
